### ISSUE TITLE
False positive in test code immediately below an assertion that the value was not null

### ISSUE THREAD

**Issue by hakanai** [uber/NullAway#501]
Example class:

```java
import javax.annotation.Nullable;
public class A {
    @Nullable
    public String getThing() {
        return "ok";
    }
}
```

Example test:

```java
import org.junit.jupiter.api.Test;
import static org.hamcrest.MatcherAssert.*;
import static org.hamcrest.Matchers.*;
public class TestA {
    @Test
    public void testGetThing() {
        String thing = new A().getThing();
        assertThat(thing, is(notNullValue()));
        assertThat(thing.length(), is(2));
    }
}
```

I get an error from NullAway at `thing.length()` even though on the previous line I had asserted that it was not null.

In this particular case, I just assume that NullAway should not be applied to test code because it doesn't understand it yet, and it's relatively easy to turn it on only for production code, so that's probably what we'll do.

Versions:
* NullAway version: 0.9.1
* ErrorProne version: 2.4.0
* Hamcrest 2.2
* JUnit 5.7.1
* Java 11

**Comment by msridhar** [comment]
Thanks for the report!  This kind of issue is handled by NullAway's library models in general.  To be honest, I think handling this case will be low priority, both because it doesn't seem like a common way to assert non-nullness, and also it would be a bit messy and require a bunch of pattern matching to do it (we actually may need a custom handler for it rather than a library model).

Are you able to find a more direct way to assert non-nullness, like an `assertNonNull` method?  We are more likely to add support for that (if we don't already support it).

**Comment by hakanai** [comment]
We moved to Hamcrest exclusively for all assertions many years ago, so going back to the older JUnit 3 style assertions is probably quite unlikely. The most we could hope for is perhaps to move the entire codebase to AssertJ style, where instead it would be `assertThat(x).isNotNull()`... but that's a lot of work to migrate the entire test suite across.

**Comment by lazaroclapp** [comment]
I believe both `assertThat(x).isNotNull()` and `assertThat(thing, is(notNullValue()))` are handled as long as you pass `-XepOpt:NullAway:HandleTestAssertionLibraries=true` to NullAway. 

I am less sure about the second idiom that I am about the first but, looking at the handler code, it looks like it is trying to match either pattern.

We could consider turning on that handler by default, actually. There is a small performance penalty, which is why internally we enable it only for test code and not for service/app code, but that's worth balancing against discoverability. Even I didn't remember we had this until the much more common (for us) `assertThat(x).isNotNull()` idiom was brought up 😅

**Comment by hakanai** [comment]
Aha, neat, I will definitely try turning that on the next time I get a chance. Yeah, there's so many levels of nesting with some Hamcrest code that it makes inspection hard. IDEA also didn't treat this case right initially.

**Comment by msridhar** [comment]
> At the very least, worth using our new performance benchmarks to double-check if the impact of turning this support on by default is something we really need to worry about or not. We could even turn it on/off depending which testing libraries are in the classpath, but that might be brittle...

I agree that (when someone gets some time) we should benchmark turning the handler on by default and see if it has any impact on our benchmarks.

**Comment by hakanai** [comment]
Turning this on makes it work for us - mostly. I'll close this and raise something more specific about the minor issue I found with it.
